NOT TO MERGE - Feature/180 support cf

arpithaar opened the pull request:
@davidjgonzalez , @godanny86 

This is an initial commit for the enhancement https://github.com/Adobe-Marketing-Cloud/asset-share-commons/issues/180.

Marking this PR "NOT TO MERGE" since the complete support for Content Fragments is not part of this pull request yet. The intent is to get your feedback/review on whatever has been done so far.

Following are the changes incorporated to support Content Fragments as of now.

1. Modified AssetTypeImpl.java to support Content Fragments
2. Modified Card and list view templates . Card view and List view displays thumbnail(if present) for content fragments. In the absence of thumbnail, CF excerpt is displayed in card view and list view is authorable to select a displayable thumbnail image for CFs
![image](https://user-images.githubusercontent.com/34283439/39831209-60923e82-53e1-11e8-8c9a-7bddece198f9.png)
3. Modified Card view template to not display resolution for Content Fragments
![image](https://user-images.githubusercontent.com/34283439/39831338-c26a1ac6-53e1-11e8-88ea-9e0968f99898.png)
4. Removed Resolution column from List view and added TYPE
![image](https://user-images.githubusercontent.com/34283439/39831422-04db48bc-53e2-11e8-964b-c7a87c18fd59.png)

Changes in content:
1. Created a Sample CF Model with Multiline Text Field under /conf/asset-share-commons/settings/dam/cfm/models
2. Created a folder at /content/dam/asset-share-commons/en/public/contentfragments with sample content fragments
3. Added details pages for content fragments - /content/asset-share-commons/en/light/details/content-fragment and /content/asset-share-commons/en/dark/details/content-fragment
4. Removed hidden filters that restricted content fragments search and listing

godanny86 commented:
Thanks @arpithaar for the PR! will have some time to review tomorrow.

davidjgonzalez commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/AssetModel.java:
```diff
@@ -62,4 +62,19 @@
+     *
+     * @return true if the asset is a Content Fragment
+     */
+    boolean isContentFragment();
```
We shouldnt add these methods to the AssetModel interface, instead we should leverage computed properties..

ie. `${asset['assetType'] == 'CONTENT-FRAGMENT'}`

Or if you really want something that returns a boolean, then a new computed property .. `${asset['isContentFragment']}` .. tho this feels overkill?

davidjgonzalez commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/AssetModel.java:
```diff
@@ -62,4 +62,19 @@
+     *
+     * @return excerpt of the Content Fragment
+     */
+    String getDisplayableExcerpt();
```
Same, this is too specific for the interface, should move to a computed property... `${asset['contentFragmentThumbnail']}` or something..

davidjgonzalez commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/AbstractComputedProperty.java:
```diff
@@ -59,23 +61,40 @@ protected ValueMap getMetadataProperties(Asset asset) {
         return metadataResource.getValueMap();
     }
 
+    protected ValueMap getJcrProperties(final Asset asset) {
```
can `getAssetProperties().get("jcr:content/someProp")` suffice?
Else maybe better name this method to `getJcrContentProperties()`? `getJcrProperties()` could mean anything - theyre all JCR properties! :)

davidjgonzalez commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/AssetTypeImpl.java:
```diff
@@ -70,16 +70,22 @@ public String getLabel() {
 
-        if (lookedupResource != null && !ResourceUtil.isNonExistingResource(lookedupResource)) {
+        /* handles content fragments */
+        if (null != assetResource.adaptTo(ContentFragment.class)) {
```
Once we sort of what computed property-based approach we can use to determining if its a CF... we can do something like [0], get that computed property and use that to perform the check. Ideally we isolate all the "is CF" checks centrally.


[0] https://github.com/Adobe-Marketing-Cloud/asset-share-commons/blob/develop/core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/ResolutionImpl.java#L43

arpithaar commented:
@davidjgonzalez 

How does it look now? Following changes have been incorporated.

1. Removed CF related APIs from AssetModel
2. Created a new Computed Property Implementation ContentFragmentExcerptImpl to display excerpt in the absence of Thumbnails
3. Made use of AssetType Computed property to determine if an asset is a content fragment
4. Introduced a Computed Property type : cf-rendition
5. Thumbnail and excerpt computed property implementations could be the possible candidates of type cf-rendition
6. A component to render CF could include the above computedPropertyTypes and master variant ?

godanny86 commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/AssetTypeImpl.java:
```diff
@@ -143,6 +149,10 @@ protected void activate(Cfg cfg) {
         )
         String audioLabel() default AUDIO_LABEL;
 
+        @AttributeDefinition(
```
I like this approach of using a computed property to make the determination "is a CF". However the configurable label is an issue here. If a user changes the label in the osgi config everything related to Content Fragments breaks.
![image](https://user-images.githubusercontent.com/8974514/40267747-411ff9da-5b30-11e8-9b75-b38a54c77986.png)

Originally the AssetType class was intended as a "display" of the high level mimetype and was not intended to do any logic based switching. The configurable labels is a nice touch but maybe its not necessary (i18n should take care of localization use case). I'm wondering if we simply do away with the configurable labels? Or we should introduce a 2nd computed property to differentiate between Type and Display Type. 

wdyt @davidjgonzalez , @arpithaar ?

godanny86 replied:
I don't love the mixing of `adaptTo` language being called as a computed property. seems dirty?
+1 for `contentFragment` -> returning a boolean.

godanny86 commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/TitleImpl.java:
```diff
@@ -52,7 +58,15 @@ public String getLabel() {
+            }
+            return jcrValueMap.get(JcrConstants.JCR_TITLE, String.class);
+        }
+
```
I think the ContentFragment API should be used when possible:

```
  if ("CONTENT-FRAGMENT".equals(assetType.get(asset))) {
            Resource res = asset.adaptTo(Resource.class);
            ContentFragment cf = res.adaptTo(ContentFragment.class);
           return cf.getTitle();
        }
```

godanny86 commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/AbstractComputedProperty.java:
```diff
@@ -59,23 +61,40 @@ protected ValueMap getMetadataProperties(Asset asset) {
         return metadataResource.getValueMap();
     }
 
+    protected ValueMap getJcrContentProperties(final Asset asset) {
```
I don't think this method is necessary. If you need to pull the `jcr:title` off the `jcr:content` node you can:

```
ValueMap vm = getAssetProperties(asset);
vm.get("jcr:content/jcr:title", String.class);
```

godanny86 commented on ui.apps/src/main/content/jcr_root/apps/asset-share-commons/components/search/results/result/card/templates/card.html:
```diff
@@ -27,9 +27,22 @@
+        	<sly data-sly-test="${asset.properties['type'] == 'CONTENT-FRAGMENT'}">
+				<img src="${asset.properties['thumbnail']}"
+				     class="cmp-image--card"
+				     alt="${asset.properties['title']}"
```
Lets break out the content region by high level asset type instead of having multiple checks for Content Fragment. Can do it via templates similar to [0] and [1] with another file called "asset-types.html". It doesn't have to be a 1:1 mapping. I think 2 templates should work for now, a default group and then a "group" for CF+Document+Presentation since Document/Presentation also do not have Resolution.

![image](https://user-images.githubusercontent.com/8974514/40268261-3a1f37f6-5b38-11e8-97a9-51ffb4680599.png)

[0] https://github.com/Adobe-Marketing-Cloud/asset-share-commons/blob/develop/ui.apps/src/main/content/jcr_root/apps/asset-share-commons/components/details/metadata/metadata.html.

[1] https://github.com/Adobe-Marketing-Cloud/asset-share-commons/blob/develop/ui.apps/src/main/content/jcr_root/apps/asset-share-commons/components/details/metadata/templates/field-types.html

arpithaar replied:
@godanny86 , @davidjgonzalez 
Wondering how do I classify assets into default group(media types) and Documents. Using the same AssetType Computed Property ? If so, I think we should simply do away with the configurable labels

godanny86 commented on ui.content/src/main/content/jcr_root/content/asset-share-commons/en/dark/details/content-fragment/.content.xml:
```diff
@@ -0,0 +1,199 @@
+    <jcr:content
+        cq:lastModified="{Date}2018-05-07T16:53:18.765+05:30"
+        cq:lastModifiedBy="admin"
+        cq:template="/conf/asset-share-commons/settings/wcm/templates/details-template"
```
should be using the Dark Template: `/conf/asset-share-commons/settings/wcm/templates/details-template-dark`

godanny86 commented on ui.apps/src/main/content/jcr_root/apps/asset-share-commons/components/search/results/clientlibs/site/less/excerpt.less:
```diff
@@ -0,0 +1,24 @@
+    		border: none;
+          	border-top: .0625rem solid #e6e6e6;
+        }
+}
```
Can the :hover color on the excerpt be removed?
![image](https://user-images.githubusercontent.com/8974514/40268334-7e0e3aec-5b39-11e8-9e5a-d998d23d8c4b.png)

arpithaar commented on core/src/main/java/com/adobe/aem/commons/assetshare/content/properties/impl/ContentFragmentExcerptImpl.java:
```diff
@@ -0,0 +1,126 @@
+    private String getCFExcerpt(final Iterator<ContentElement> elements, final int maxChars) {
+        String excerpt = "";
+
+        while (elements.hasNext() && excerpt.length() < maxChars) {
```
@davidjgonzalez ,
I guess Sling HTML Parser doesn't provide APIs to remove html tags from a String. I would like to use convertToType api of ContentTypeConverter(com.adobe.cq.dam.cfm.converter.ContentTypeConverter). Thoughts?

arpithaar commented:
Closing this. Shall raise a new PR
